#180 Add Attach classes to JNI file

jperedadnr opened the pull request.

johanvos commented on src/main/java/com/gluonhq/substrate/target/AbstractTargetConfiguration.java:
```diff
@@ -636,6 +650,11 @@ String getAppPath(String appName) {
         return Collections.emptyList();
     }
 
+    // by default, false, but subclasses can decide against it.
```
What is this method supposed to do/return?

jperedadnr replied:
I've updated the javadoc

johanvos commented on src/main/java/com/gluonhq/substrate/config/ConfigResolver.java:
```diff
@@ -0,0 +1,149 @@
+        Logger.logDebug("Scanning for init build time files");
+        return scanJars(USER_INIT_BUILD_TIME_FILE,
+                Strings.substitute(USER_INIT_BUILD_TIME_ARCHOS_FILE, Map.of("archOs", archOs)),
+                null,
```
looks like we can use a Strings.substitute method with less parameters?

eugener replied:
@johanvos Not sure what you mean. I don't think we have a shorter one.

johanvos replied:
then why don't we add a Strings.substitute() method with less parameters? If it is legal to call foo(A, B) with B null, I think it makes sense to add foo(A) that will default to foo(A, null). That looks cleaner to me than the caller passing null as an argument.

eugener replied:
I think `null` is passed into `scanJars`, not into `substitute`?

johanvos replied:
you're right. So that means my remark is about scanJars :)

eugener commented on src/main/java/com/gluonhq/substrate/config/ConfigResolver.java:
```diff
@@ -0,0 +1,149 @@
+     * @throws IOException
+     */
+    public List<String> getUserInitBuildTimeList(String archOs) throws IOException {
+        Logger.logDebug("Scanning for init build time files");
```
Would be nice to check the `archOS` for null. Previous method does not need this as `Classpath` does it already.

eugener commented on src/main/java/com/gluonhq/substrate/config/ConfigResolver.java:
```diff
@@ -0,0 +1,149 @@
+     * @throws IOException
+     */
+    public List<String> getUserReflectionList(String archOs) throws IOException {
+        Logger.logDebug("Scanning for reflection files");
```
Check the argument for null

eugener commented on src/main/java/com/gluonhq/substrate/config/ConfigResolver.java:
```diff
@@ -0,0 +1,149 @@
+     * @throws IOException
+     */
+    public List<String> getUserJNIList(String archOs) throws IOException {
+        Logger.logDebug("Scanning for JNI files");
```
Check the argument for null

eugener commented on src/main/java/com/gluonhq/substrate/config/ConfigResolver.java:
```diff
@@ -0,0 +1,149 @@
+    private final List<File> jars;
+
+    public ConfigResolver(String classpath) {
+        this.jars = new ClassPath(classpath)
```
Since we require class path as a constructor parameter, why not to use `ClassPath`? 
I suspect that that is what the caller would have anyway?

jperedadnr replied:
Right, but this probably leads to a bigger refactoring. We have `processClassPath`, that returns a String, while it probably could return a `ClassPath` object, and that will be the only object we would need to create:
```
public boolean compile(String cp) throws IOException, InterruptedException {
        ClassPath classPath = processClassPath(cp);
    ...
```

Does that make sense? Probably for a follow-up PR then.

eugener commented on src/main/java/com/gluonhq/substrate/config/ConfigResolver.java:
```diff
@@ -0,0 +1,149 @@
+    }
+
+    private List<String> scanJars(String configName, String configArchosName, String initLine, Predicate<String> filter) throws IOException {
+        List<String> list = new ArrayList<>();
```
Check arguments for nulls

jperedadnr commented:
`ConfigResolver` depends on the classpath and `scanJars` will process all found jars in it. 
In order to provide some tests, I see two options: 
- add a dummy jar with some dummy content in `META-INF/substrate/config` to resources, then before the test, copy this jar in some tmp folder, and pass that path as "classpath", so we can process it and check the results.
- test-project/HelloGluon will use it in a real case scenario, once GM and Attach are released including the `META-INF/substrate/config` real content.

jperedadnr commented:
I've added unit tests, based on a dummy jar I've created and added to test/resources. 
@eugener Do you think it is necessary to keep the sources of that jar and how to generate it? It could be automated, but I don't think it is worth the hassle?

eugener commented:
I don’t think source of the test jar is that important as it will hopefully never change.

eugener commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/main/java/com/gluonhq/substrate/config/ConfigResolver.java  8
- src/test/java/com/gluonhq/substrate/config/ConfigTests.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/gluonhq/substrate/pullRequest?prid=4678876&bid=15652398)

eugener commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/test/java/com/gluonhq/substrate/config/ConfigTests.java  1
- src/main/java/com/gluonhq/substrate/config/ConfigResolver.java  8
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/gluonhq/substrate/pullRequest?prid=4678876&bid=15652398)
